### PR TITLE
Integrated PR: PR of multiple PRs ready to be merged

### PR DESCRIPTION
**Pull request by prithviraj-gotepatil-algobulls**
This is a Integrated PR. Multiples PRs are merged in this PR.
Following are the PR numbers of merged PRs.
#32 
#34 
#41 
#55 
#59 
#60 
#61 
#62 
#67

### CODE REVIEW

File: pyalgotrading/algobulls/api.py [hunk 1]
```diff
@@ -335,7 +335,7 @@ def start_strategy_algotrading(self, strategy_code: str, start_timestamp: dt, en
                 TradingType.BACKTESTING: 'backDataDate'
             }
             execute_config = {
-                map_trading_type_to_date_key[trading_type]: [start_timestamp.astimezone(timezone.utc).isoformat(), end_timestamp.astimezone(timezone.utc).isoformat()],
+                map_trading_type_to_date_key[trading_type]: [start_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z"), end_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z")],
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/api.py:338]: Try this for the conversion - 
start_timestamp.astimezone(timezone.utc).replace(tzinfo=None).isoformat()

Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/api.py]: Done

File: pyalgotrading/algobulls/connection.py [hunk 2]
```diff
@@ -398,26 +443,89 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
         assert isinstance(render_as_dataframe, bool), f'Argument "render_as_dataframe" should be a bool'
         assert isinstance(show_all_rows, bool), f'Argument "show_all_rows" should be a bool'
         # assert (broker is None or isinstance(broker, AlgoBullsSupportedBrokers) is True), f'Argument broker should be None or an enum of type {AlgoBullsSupportedBrokers.__name__}'
-        response = self.api.get_reports(strategy_code=strategy_code, trading_type=trading_type, report_type=report_type, country=country)
-        if response.get('data'):
+
+        if report_type is TradingReportType.PNL_TABLE:
+            response = self.api.get_reports(strategy_code=strategy_code, trading_type=trading_type, report_type=report_type, country=country, current_page=1)
+            main_data = response.get("data")
+        else:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:450]: what report types are part of the else? Both stats and order history?

File: pyalgotrading/algobulls/connection.py [hunk 3]
```diff
@@ -398,26 +443,89 @@ def get_report(self, strategy_code, trading_type, report_type, render_as_datafra
...
+
+            while True:
+                response = self.api.get_reports(strategy_code=strategy_code, trading_type=trading_type, report_type=report_type, country=country, current_page=i)
+                _total = response.get("totalTrades")
+                _data = response.get("data")
+
+                # if data is retrieved as a list then update the main data
+                if _data and isinstance(_data, list):
+                    main_data.extend(_data)
+                    total_data += 1000
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:463]: can the hardcoding be avoided?

File: pyalgotrading/algobulls/connection.py [hunk 4]
```diff
@@ -455,13 +563,33 @@ def get_pnl_report_table(self, strategy_code, trading_type, country):
...
+                    _df['entry_variety'] = 'MARKET'
+                    print('WARNING: Column for Order Variety not found. Assuming all trades are Market Orders.')
+
+                _df[['entry_price', 'exit_price']] = _df.apply(
+                    lambda row: (slippage(row.entry_price, row.entry_variety, row.entry_transaction_type, slippage_percent), slippage(row.exit_price, row.exit_variety, row.exit_transaction_type, slippage_percent)), axis=1, result_type='expand')
+
+                _df['pnl_absolute'] = _df['exit_price'] - _df['entry_price']
+
+            # add brokerage
+            _df['brokerage'] = ((_df['entry_price'] * _df['entry_quantity']) + (_df['exit_price'] * _df['exit_quantity'])) * (broker_commission_percentage / 100)
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:579]: There should be a way to turn off brokerage computation. Don't see it right now.

File: pyalgotrading/algobulls/connection.py [hunk 5]
```diff
@@ -972,24 +1145,32 @@ def get_papertrading_report_statistics(self, strategy_code, initial_funds=1e9, m
         else:
             print('Generating Statistics for already fetched P&L data...')
 
+        if initial_funds is None:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:1148]: can we move this logic inside get_report_statistics? will it break anything?

Comment (pushpakdagade) [reply pyalgotrading/algobulls/connection.py:1148]: if no, let's do it at other places with similar logic.

File: pyalgotrading/algobulls/connection.py [hunk 6]
```diff
@@ -972,24 +1145,32 @@ def get_papertrading_report_statistics(self, strategy_code, initial_funds=1e9, m
...
+            render_as_dataframe: return order history as dataframe or pretty string
 
         Returns:
             Report details
         """
 
         assert isinstance(strategy_code, str), f'Argument "strategy_code" should be a string'
 
-        return self.get_report(strategy_code=strategy_code, trading_type=TradingType.PAPERTRADING, report_type=TradingReportType.ORDER_HISTORY)
+        if country is None:
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:1170]: can we move this logic inside get_report? will it break anything? if no, let's do it at other places with similar logic

File: pyalgotrading/algobulls/connection.py [hunk 7]
```diff
@@ -972,24 +1145,32 @@ def get_papertrading_report_statistics(self, strategy_code, initial_funds=1e9, m
...
+            country = self.strategy_country_map[TradingType.PAPERTRADING].get(strategy_code, Country.DEFAULT.value)
+
+        return self.get_report(strategy_code=strategy_code, trading_type=TradingType.PAPERTRADING, report_type=TradingReportType.ORDER_HISTORY, render_as_dataframe=render_as_dataframe, country=country)
```
Comment (pushpakdagade) [inline pyalgotrading/algobulls/connection.py:1173]: should we have a dedicated function for getting order history report like we have for statistics, get_report_statistics?

File: pyalgotrading/strategy/__init__.py [hunk 8]
```diff
@@ -1 +1,3 @@
+from ..constants import *
```
Comment (pushpakdagade) [inline pyalgotrading/strategy/__init__.py:1]: is this required?

File: pyalgotrading/algobulls/connection.py (hunk 2 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py:450]: `PNL_TABLE` and `ORDER_HISTORY` are the 2 report types available here

File: pyalgotrading/algobulls/connection.py (hunk 4 above)
Comment (prithviraj-gotepatil-algobulls) [reply pyalgotrading/algobulls/connection.py:579]: If brokerage is not given then it will give 0 brokerage.

**Pull request merged**